pkg/destroy/vsphere/virtualmachines.go

```diff
@@ -0,0 +1,128 @@
+package vsphere
+
+import (
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+func (o *ClusterUninstaller) getVirtualMachineManagedObjects(moRef []types.ManagedObjectReference) ([]mo.VirtualMachine, error) {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	var virtualMachineMoList []mo.VirtualMachine
+	pc := property.DefaultCollector(o.Client)
+	err := pc.Retrieve(ctx, moRef, nil, &virtualMachineMoList)
+	if err != nil {
+		return nil, err
+	}
+	return virtualMachineMoList, nil
+}
+
+func (o *ClusterUninstaller) listVirtualMachines() ([]mo.VirtualMachine, error) {
+	tagAttachedObjects, err := o.getAttachedObjectsOnTag()
+	if err != nil && !isNotFound(err) {
+		return nil, err
+	}
+
+	// Seperate the objects attached to the tag based on type
+	// We only need VirtualMachine
+	var virtualMachineList []types.ManagedObjectReference
+	for _, attachedObject := range tagAttachedObjects {
+		for _, ref := range attachedObject.ObjectIDs {
+			if ref.Reference().Type == "VirtualMachine" {
+				virtualMachineList = append(virtualMachineList, ref.Reference())
+			}
+		}
+	}
+
+	var virtualMachineMoList []mo.VirtualMachine
+	if len(virtualMachineList) > 0 {
+		virtualMachineMoList, err = o.getVirtualMachineManagedObjects(virtualMachineList)
+	}
+
+	return virtualMachineMoList, err
+}
+
+func isPoweredOff(vmMO mo.VirtualMachine) bool {
+	return vmMO.Summary.Runtime.PowerState == "poweredOff"
+}
+
+func (o *ClusterUninstaller) stopVirtualMachine(vmMO mo.VirtualMachine) error {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	virtualMachineLogger := o.Logger.WithField("VirtualMachine", vmMO.Name)
+	if !isPoweredOff(vmMO) {
+		vm := object.NewVirtualMachine(o.Client, vmMO.Reference())
+		task, err := vm.PowerOff(ctx)
+		if err == nil {
+			err = task.Wait(ctx)
+		}
+		if err != nil {
+			virtualMachineLogger.Debug(err)
+			return err
+		}
+	}
+	virtualMachineLogger.Debug("Powered off")
+	return nil
+}
+
+func (o *ClusterUninstaller) stopVirtualMachines() error {
+	o.Logger.Debug("Power Off Virtual Machines")
+	found, err := o.listVirtualMachines()
+	if err != nil {
+		o.Logger.Debug(err)
+		return err
+	}
+
+	var errs []error
+	for _, vmMO := range found {
+		if !isPoweredOff(vmMO) {
+			if err := o.stopVirtualMachine(vmMO); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (o *ClusterUninstaller) deleteVirtualMachine(vmMO mo.VirtualMachine) error {
+	ctx, cancel := o.contextWithTimeout()
```

Comment:
Is the 2-minute timeout enough? We were using a 30-minute timeout for the entire operation of powering off and destroying all of the VMs.

Reply:
I can increase the `defaultTimeout` from 2min to something else (I was just using the same value as in GCP) or I can use a non-default here. Just let me know which timeout value you'd prefer.

Reply:
I think GCP is different because the requests made in GCP do not wait for long-running operations to complete before returning. The requests return when the operation has started and the state of the resource has been changed to indicate that. It appears that the vSphere API requests wait until the long-running operation completes before returning.